File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silent?(input) -> "Fine. Be that way!"
      question?(input) -> "Sure."
      shouting?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp silent?(input) do
    String.strip(input) == ""
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  defp shouting?(input) do
    String.upcase(input) == input and has_letter?(input)
  end

  defp has_letter?(input) do
    String.match?(input, ~r/\p{L}/)
  end
end

File: go/bob/bob.go
package bob

import "strings"
import "regexp"

func Hey(remark string) string {
  trimmed_remark := strings.TrimSpace(remark); 
  if shouting(trimmed_remark) && has_letter(trimmed_remark) {
    return "Whoa, chill out!"
  } else if strings.HasSuffix(trimmed_remark, "?") {
    return "Sure."
  } else if len(trimmed_remark) == 0 {
    return "Fine. Be that way!"
  }

  return "Whatever."
}

func shouting(remark string) bool {
  return strings.ToUpper(remark) == remark
}

func has_letter(remark string) bool {
  matched, _ := regexp.Match("[A-Z]", []byte(remark))

  return matched
}

func question(remark string) bool {
  return strings.HasSuffix(remark, "?")
}
